### ISSUE TITLE
Patch log level config on runtime

### ISSUE THREAD

**Issue by alileza** [prometheus/prometheus#10352]
## Proposal
**Use case. Why is this important?**

In some cases we would like to know what's going on with prometheus in production, we would like to see prometheus log change to debug. I thought it would be nice to have a way to change the log level in runtime without interrupting prometheus work (such as deployment).

So I'm thinking something like adding `PATCH /config` endpoint to prometheus could be nice, or maybe just specific only for the log level

**Comment by roidelapluie** [comment]
Yes, that's something I would like to add to Prometheus. it would involve moving the log level to file. this should be discussed in the dev summit because I don't know if everyone is happy with deprecating the command line flag.

**Comment by krajorama** [comment]
Hello from the bug scrub.

This was not yet discussed at the dev summit, we're adding it to the agend for the next in-person dev summit.

**Comment by beorn7** [comment]
Update from the dev summit:

We finally discussed this. We agreed to introduce a config option (which will implicitly allow changes of the log level at runtime because the config is reloadable). This also implies to deprecate the command line flag, i.e. if you use it, you'll see a deprecation warning, and we will remove the flag in v4 (which might be years in the future).

The dev summit was very aware that there might be resistance against removing the command line option because it's very common to configure the log level via the command line. On the other hand, we also felt that breaking the well established patterns would also be very problematic (which is to have runtime changeable options in the config file and static options on the command line and never have two ways of configuring the same thing). So we settled on doing what's stated above, but of course, we can still reconsider in time before v4 is released.

**Comment by bwplotka** [comment]
I might need to implement this for easier Prometheus debugging. I plan to start by introducing the config option. The CLI setting will control the logging from start to config parsing AND the default option for config parser. 

Personally. I think this would be ok to leave the CLI in such semantics, but if the group wants to deprecate it, can do.

**Comment by loglapa** [comment]
I'd like to implement the direction agreed at the dev summit: a reloadable log-level config option, with the CLI flag supplying the startup/default value and a deprecation warning rather than being removed. I'll add tests proving a successful reload changes filtering live and a failed reload preserves the previous level, plus the relevant CLI/config documentation. I'll keep the PR scoped to log-level lifecycle and compatibility.
